File: 2º Período/tp03/ListaFlex.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

#define MAXTAM 3923
typedef struct Jogador
{
		int id;
		char nome[100];
		int altura;
		int peso;
		char universidade[100];
		int anoNascimento;
		char cidadeNascimento[100];
		char estadoNascimento[100];
} Jogador;
typedef struct Celula
{
	struct Celula* prox;
	struct Celula* ant;
	Jogador jogador;
}Celula;
Celula* novaCelula(Jogador x)
{
	Celula *tmp = (Celula*) malloc(sizeof(Celula));
	tmp->jogador = x;
	tmp->prox = NULL;
	tmp->ant = NULL;
	return tmp;
}
 typedef struct Lista
{
	int tamanho;
	struct Celula* primeiro;
	struct Celula* ultimo;
}Lista;
 Lista NovaLista()
{
   Lista tmp;
   Jogador a;
   tmp.primeiro = novaCelula(a);
   tmp.ultimo = tmp.primeiro; 
   tmp.tamanho = 0;
   return tmp;
}

 int getId(Jogador* x)
 {
	return x->id;
 }
 char* getNome(Jogador* x )
{
	return x->nome;
}
 int getAltura(Jogador* x)
{
	return x->altura;
}
 int getPeso(Jogador* x)
{
	return x->peso;
}
 char* getUniversidade(Jogador* x)
{
	return x->universidade;
}
 int getAnoNascimento(Jogador* x)
{
	return x->anoNascimento;
}
 char* getCidadeNascimento(Jogador* x)
{
	return x->cidadeNascimento;
}
 char* getEstadoNascimento(Jogador* x)
{
	return x->estadoNascimento;
}
void setId(Jogador *x, int novoId)
{
	x->id = novoId;
}
void setNome(Jogador* x, char* nome)
{
	strcpy(x->nome, nome);
}
void setAltura(Jogador* x, int altura)
{
	x->altura = altura;
}
void setPeso(Jogador* x, int peso)
{
	x->peso = peso;
}
void setUniversidade(Jogador *x, char* universidade)
{
	strcpy(x->universidade, universidade);
}
void setAnoNascimento(Jogador* x, int anoNascimento)
{
	x->anoNascimento = anoNascimento;
}
void setCidadeNascimento(Jogador* x, char* cidadeNascimento)
{
	strcpy(x->cidadeNascimento, cidadeNascimento);
}
void setEstadoNascimento(Jogador* x, char* estadoNascimento)
{
	strcpy(x->estadoNascimento, estadoNascimento);
}
//metodo para criar um clone do jogador
Jogador clone(Jogador *x)
{
	Jogador clone;
	clone.id = getId(x);
	strcpy(clone.nome, getNome(x));
	clone.altura = getAltura(x);
        clone.peso = getPeso(x);
        clone.anoNascimento = getAnoNascimento(x);
        strcpy(clone.universidade, getUniversidade(x));
	strcpy(clone.cidadeNascimento, getCidadeNascimento(x));
	strcpy(clone.estadoNascimento, getEstadoNascimento(x));

	return clone;
}
//Metodo para ler o arquivo e retornar um ponteiro para os Jogadores selecionados
Jogador* ler()
{
 FILE* arq = fopen("/tmp/players.csv","r");
 if(arq == NULL)
 {
	 printf("nao abriu");
 }
 char temp[1000];
 int cont;
 int num = 0;
 Jogador *y=malloc(sizeof(Jogador));
 Jogador *resp = malloc(3923*sizeof(Jogador));
   while(fscanf(arq, "%[^\n]", temp) != EOF)
   {
	   cont = 0; 
	   fgetc(arq);
	   char *items[8];
	   items[4]="";
	   items[6]="";
	   items[7]="";
           char *token  = strtok(temp,",");
	   for(int i = 0; token ; i++)
	   {
            items[i] = token;
            token = strtok(NULL, ",");
            cont++;
	   }
//ifs para tratar casos em que existiam informações faltantes
	   if(cont == 7 && strlen(items[7])==0)
       {
           items[7] = items[6];
       }
	   else if(cont ==5)
	   {
		   items[5] = items[4];
		   items[4] = "";
	   }
       if(atoi(items[5]) == 0)
       {
           items[6] = items [5];
       }
	   if(atoi(items[4]) != 0)
       {
           items[5] = items[4];
       }

	   setId(y, atoi(items[0]));
	   setNome(y, items[1]);
	   setAltura(y, atoi(items[2]));
	   setPeso(y, atoi(items[3]));
	   setUniversidade(y, items[4]);
	   setAnoNascimento(y, atoi(items[5]));
       setCidadeNascimento(y, items[6]);
	   setEstadoNascimento(y, items[7]);

//ifs para setar corretamente campos não informados
           if(strlen(items[4])==0 || atoi(getUniversidade(y))>0)
	   {
	           setUniversidade(y, "nao informado");
	   }
	     if(strlen(items[6])==0)
	   {
		  setCidadeNascimento(y, "nao informado");
	   }
	     if(strlen(items[7])==0)
	   {
		   setEstadoNascimento(y, "nao informado");
	   }
	     if(getAnoNascimento(y)==0)
         {
             setAnoNascimento(y, atoi(items[4]));
         }
         
//comparar o input do usuario com o que esta sendo lid
	 resp[num] =clone(y);
	 num++;
   }
   fclose(arq);
   return resp;
}

//função para imprimir o jogador selecionado
void imprimir(Lista *l)
{
	int j = 0;
	for(Celula* i =l->primeiro->prox; i!=NULL; i = i->prox)
	{
	printf("[");
	printf("%d", getId(&i->jogador));
	printf(" ## ");
	printf("%s", getNome(&i->jogador));
	printf(" ## ");
	printf("%d", getAltura(&i->jogador));
	printf(" ## ");
	printf("%d", getPeso(&i->jogador));
	printf(" ## ");
	printf("%d", getAnoNascimento(&i->jogador));
	printf(" ## "); 
	printf("%s", getUniversidade(&i->jogador));
	printf(" ## "); 
	printf("%s", getCidadeNascimento(&i->jogador));
	printf(" ## ");
	printf("%s]", getEstadoNascimento(&i->jogador));
	printf("\n");
	}
}
//metodo para trocar dois jogadores de posição no array
void swap(Jogador *i, Jogador *j)
{
	Jogador tmp = *i;
	*i = *j;
	*j = tmp;
	
}
//metodo para comparar jogadores com o atributo primário sendo o EstadoNascimento e o secundário o nome
 int compare2(Jogador *x, Jogador *y) 
{
    char jog1[200];
    char jog2[200];
    strcpy(jog1, getEstadoNascimento(x));
    strcat(jog1, getNome(x));    
    strcpy(jog2, getEstadoNascimento(y));
    strcat(jog2, getNome(y));
      return strcmp(jog1, jog2);
}
void inserirInicio(Jogador x, Lista *l)
{
	Celula *tmp = novaCelula(x);
	tmp->prox = l->primeiro->prox;
	l->primeiro->prox = tmp;
	tmp->ant = l->primeiro;
	if(l->ultimo == l->primeiro)
		l->ultimo = tmp;
	else
		tmp->prox->ant = tmp;
	tmp = NULL;
	l->tamanho++;
	
}
void inserirFim(Jogador x, Lista *l)
{
		l->ultimo->prox = novaCelula(x);
		l->ultimo->prox->ant = l->ultimo;
		l->ultimo = l->ultimo->prox;
		l->tamanho++;
}
void inserir(int pos, Jogador x, Lista *l)
{
	if(pos <0 || pos> l->tamanho)
		return;

	else if( pos == 0)
	{
		inserirInicio(x, l);
	}
	else if(pos == l->tamanho)
	{
		inserirFim(x, l);
	}
	else
	{
		Celula *i = l->primeiro;
		for(int x = 0; x < pos; x++)
		{
			i = i->prox;

		}
		Celula *tmp = novaCelula(x);
		tmp->prox = i->prox;
		tmp->ant = i;
		i->prox->ant = tmp;
		i->prox = tmp;
		i = tmp = NULL;
		l->tamanho++;
	}


}
Jogador removerInicio(Lista *l)
{
       if(l->tamanho == 0)
       {
	       printf("Array vazio");
	       exit(1);
       }
       Celula *tmp = l->primeiro->prox;
       Jogador removido = tmp->jogador;
       l->primeiro->prox = tmp->prox;
       tmp->prox->ant = l->primeiro;
       free(tmp);
       tmp = NULL;
       l->tamanho--;
       return removido;
}
Jogador removerFim(Lista *l)
{
	if(l->tamanho == 0)
	{
		printf("Array vazio");
		exit(1);
	}
        Jogador removido =  l->ultimo->jogador;
	l->ultimo = l->ultimo->ant;
	l->ultimo->prox->ant = NULL;
	free(l->ultimo->prox);
	l->ultimo->prox = NULL;
	l->tamanho--;
	return removido;

}
Jogador remover(int pos, Lista *l)
{
	Jogador removido;
	if(pos<0 || pos >= l->tamanho-1)
	{
		printf("Posição invalida");
		exit(1);
	}

	else if(pos == 0)
	{
		removido = removerInicio(l);
	}
	else if(pos == l->tamanho-1)
	{
		removido = removerFim(l);
	}
	else
	{
		Celula *i = l->primeiro->prox;
		for(int x = 0; x<pos; x++)
		{
			i = i->prox;
		}
		i->ant->prox = i->prox;
		i->prox->ant = i->ant;
		removido = i->jogador;
		i->prox = i->ant = NULL;
		free(i);
		i = NULL;
		l->tamanho--;

	}
	return removido;
}
	/**
	 * Algoritmo de ordenacao Quicksort.
    * @param int esq inicio do array a ser ordenado
    * @param int dir fim do array a ser ordenado
	 */
         void quicksort(Celula *esq, Celula *dir, Lista *l) {
        Celula *i = esq;
	Celula *j = dir;
        Celula *pivo = esq;
        while (i->ant != j && j->prox != i->ant) {
            while (compare2(&i->jogador, &pivo->jogador)<0) i = i->prox;
            while (compare2(&j->jogador, &pivo->jogador)>0) j = j->ant;
            if (i->ant != j && j->prox != i->ant ) {
                Jogador tmp = i->jogador;
		i->jogador = j->jogador;
		j->jogador = tmp;
                i = i->prox;
                j = j->ant;
            }
        }
	if (esq != j && esq->ant != j)  quicksort(esq, j, l);
	if (i != dir && i!= dir->prox)  quicksort(i, dir, l);
    } 
   void sort(Lista *l)
{
	quicksort(l->primeiro->prox, l->ultimo, l);
}
int main()
{ 
//criação de variavéis e alocação de memória
  char temp[100];
  int i =0;
  Lista l = NovaLista();
  Jogador *y = malloc(3923*sizeof(Jogador));
  y = ler();
  //while para leitura do input ate ser digitado FIM
  while(1)
  {
  scanf("%[^\n]", temp);
  char ch = getchar();
   if(strcmp(temp, "FIM")== 0)
   {
		  break; 
   }
   int num = atoi(temp);
   inserirFim(y[num+1], &l); 
  }
          sort(&l);
	 imprimir(&l);
  
 // imprimir(&l);
  free(y);
  return 0;
}

File: 2º Período/AEDS II/tp03/ListaFlex.java
class Jogador
{
    private int id;
    private String nome;
    private int altura;
    private int peso;
    private String universidade;
    private int anoNascimento;
    private String cidadeNascimento;
    private String estadoNascimento;
// Construtor sem parâmetros
    public Jogador()
    {

    }
// Construtor com parâmetros para todos atributos
    public Jogador(int id, String nome, int altura, int peso, String universidade, int anoNascimento, String cidadeNascimento, String estadoNascimento)
    {
         this.id = id;
	 this.nome = nome;
	 this.altura = altura;
	 this.peso = peso;
	 this.universidade = universidade;
	 this.anoNascimento = anoNascimento;
	 this.cidadeNascimento = cidadeNascimento;
	 this.estadoNascimento = estadoNascimento;
    }
// Metódos setters
    void setId(int id)
    {
	    this.id =id;
    }
    void setNome(String nome)
    {
	    this.nome= nome;
    }
    void setAltura(int altura)
    {
	    this.altura = altura;
    }
    void setPeso(int peso)
    {
	    this.peso = peso;
    }
    void setUniversidade(String universidade)
    {
	    this.universidade = universidade;
    }
    void setAnoNascimento(int anoNascimento)
    {
	    this.anoNascimento = anoNascimento;
    }
    void setCidadeNascimento(String cidadeNascimento)
    {
	    this.cidadeNascimento = cidadeNascimento;
    } 
    void setEstadoNascimento(String estadoNascimento)
    {
	    this.estadoNascimento = estadoNascimento;
    }
// Metódos getters
    int getId()
    {
	    return id;
    }
    String getNome()
    {
	    return nome;
    }
    int getAltura()
    {
	    return altura;
    }
    int getPeso()
    {
	    return peso;
    }
    String getUniversidade()
    {
	    return universidade;
    }
    int getAnoNascimento()
    {
	    return anoNascimento;
    }
    String getCidadeNascimento()
    {
	    return cidadeNascimento;
    }
    String getEstadoNascimento()
    {
	    return estadoNascimento;
    }
//Metodo para encontrar o jogador com o id inserido
    static Jogador PesquisaId(int x)
    {
	    Jogador[] temp = ler();
	    for(int i = 0; i<3922; i++)
	    {
		    if(temp[i].getId()==x)
		    {
			    return temp[i];
		    }
	    }
	    return temp[0];
    }
// Método para ler o jogador a partir do ID do parâmetro
    static Jogador[] ler()
    {
	    Arq.openRead("/tmp/players.csv");
	    Jogador[] jogador= new Jogador[3922];
	    int i = 0;
	    Arq.readLine();
	    while(Arq.hasNext())
	    {
            jogador[i] = new Jogador(); 
	    String jog = Arq.readLine();
	    String[] temp = new String[8];
            temp= jog.split(",", 8);
	    jogador[i].setId(Integer.parseInt(temp[0]));
            jogador[i].setNome(temp[1]);
	    jogador[i].setAltura(Integer.parseInt(temp[2]));
	    jogador[i].setPeso(Integer.parseInt(temp[3]));
	    jogador[i].setUniversidade(temp[4]);
	    jogador[i].setAnoNascimento(Integer.parseInt(temp[5]));
	    jogador[i].setCidadeNascimento(temp[6]);
	    jogador[i].setEstadoNascimento(temp[7]);

// Ifs para lidar com campos vazios
	    if(temp[4].isEmpty())
	           jogador[i].setUniversidade("nao informado");  
	    if(temp[6].isEmpty())
		   jogador[i].setCidadeNascimento("nao informado");
	    if(temp[7].isEmpty())
		   jogador[i].setEstadoNascimento("nao informado");
	    i++;
            }
	    Arq.close();
	    return jogador;


    }
// Metodo para clonar um jogador
   static Jogador clonej(Jogador jogador) {
	    String temp;
	    Jogador clone = new Jogador();
	    temp =  Integer.toString(jogador.getId());
	    clone.setId(Integer.parseInt(temp));
	    temp = jogador.getNome();
	    clone.setNome(temp);
	    temp = Integer.toString(jogador.getAltura());
	    clone.setAltura(Integer.parseInt(temp));
	    temp = Integer.toString(jogador.getPeso());
	    clone.setPeso(Integer.parseInt(temp));
	    temp = jogador.getUniversidade();
	    clone.setUniversidade(temp);
	    temp = Integer.toString(jogador.getAnoNascimento());
	    clone.setAnoNascimento(Integer.parseInt(temp));
	    temp = jogador.getCidadeNascimento();
	    clone.setCidadeNascimento(temp);
	    temp = jogador.getEstadoNascimento();
	    clone.setEstadoNascimento(temp);
	    return clone;
    }
}
class Celula
{
	protected Jogador jogador;
	protected Celula prox, ant;
	Celula()
	{
		this(null);
	}
	Celula(Jogador jogador)
	{
		this.jogador = jogador;
		prox = null;
		ant = null;
	}
}
class ListaFlex extends Jogador 
{
   private Celula ultimo,primeiro;
   private int tamanho;
   ListaFlex()
   {
	   ultimo = primeiro = new Celula();
	   tamanho = 0;
   }
  void inserirFim(Jogador x)
  {
	  ultimo.prox = new Celula(x);
	  ultimo.prox.ant = ultimo;
	  ultimo = ultimo.prox;
	  tamanho++;
  }
  void inserirInicio(Jogador x)
  {
	  Celula tmp = new Celula(x);
	  tmp.prox = primeiro.prox;
	  primeiro.prox = tmp;
	  tmp.ant = primeiro;
	  if(primeiro == ultimo)
	  {
		  ultimo = tmp;
	  }
	  else
	  {
		  tmp.prox.ant = tmp;
	  }
	  tmp = null;
	  tamanho++;
  }
  void inserir(int pos, Jogador x)
  {
	  if(pos<0 || pos > tamanho)
	  {
		  return;
	  }
	  else if(pos == 0)
	  {
		  inserirInicio(x);
	  }
	  else if(pos == tamanho)
	  {
		  inserirFim(x);
	  }
	  else
	  {
		  Celula tmp = new Celula(x);
		  Celula i = primeiro;
		  for(int j = 0; j<pos; j++)
		  {
			  i = i.prox;

		  }
		  tmp.prox = i.prox;
		  tmp.prox.ant = tmp;
		  i.prox = tmp;
		  tmp.ant = i;
		  i = null;
		  tmp = null;
		  tamanho++;
	  }
	  
  }
  Jogador removerInicio()
  {
	  if(primeiro == ultimo)
	  {
		  return null;
	  }
	  Jogador elemento = primeiro.prox.jogador;
	  Celula tmp = primeiro.prox;
	  tmp.prox.ant = primeiro;
	  primeiro.prox = tmp.prox;
	  tmp.ant =null;
	  tmp.prox = null;
	  tmp = null;
	  tamanho--;
          return elemento;

  }
  Jogador removerFim()
  {
	  if(primeiro == ultimo)
	  {
		  return null;
	  }
	  Jogador elemento = ultimo.jogador;
	  ultimo = ultimo.ant;
	  ultimo.prox.ant = null;
	  ultimo.prox = null;
	  tamanho--;
	  return elemento;
  }
  Jogador remover(int pos)
  {
	  Jogador elemento;
	  if(pos<0 || pos >= tamanho || ultimo == primeiro)
	  {
		  return null;
	  }
	  else if (pos == 0)
	  {
               elemento = removerInicio();
	  }
	  else if(pos == tamanho -1)
	  {
		  elemento = removerFim();
	  }
	  else
	  {
		  Celula i = primeiro.prox;
		  for(int j = 0; j<pos; j++)
		  {
			  i = i.prox;
		  }
		  elemento = i.jogador;
		  i.ant.prox = i.prox;
		  i.prox.ant = i.ant;
		  i.prox = i.ant = null;
		  i = null;
	          tamanho--;	  
	  }


	  return elemento;
  }

  void mostrar()
{
	int j = 0;
	for(Celula i = primeiro.prox; i!=null; i = i.prox)
	{


	    System.out.print("["+i.jogador.getId());
	    System.out.print(" ## ");
	    System.out.print(i.jogador.getNome());
	    System.out.print(" ## ");
	    System.out.print(i.jogador.getAltura());
	    System.out.print(" ## ");
	    System.out.print(i.jogador.getPeso());
	    System.out.print(" ## ");
	    System.out.print(i.jogador.getAnoNascimento());
	    System.out.print(" ## ");
	    System.out.print(i.jogador.getUniversidade());
	    System.out.print(" ## ");
	    System.out.print(i.jogador.getCidadeNascimento());
	    System.out.print(" ## ");
	    System.out.print(i.jogador.getEstadoNascimento());
	    System.out.println("]");
	}
}
  public void sort() {
      quicksort(primeiro.prox, ultimo);
   }

	/**
	 * Algoritmo de ordenacao Quicksort.
    * @param int esq inicio do array a ser ordenado
    * @param int dir fim do array a ser ordenado
	 */
      private void quicksort(Celula esq, Celula dir) {
        Celula i = esq, j = dir;
        Celula pivo = dir;
        while (i.ant != j && j.prox != i.ant) {
            while (compare2(i.jogador, pivo.jogador)<0) i = i.prox;
            while (compare2(j.jogador, pivo.jogador)>0) j = j.ant;
            if (i.ant != j && j.prox != i.ant ) {
                Jogador tmp = i.jogador;
		i.jogador = j.jogador;
		j.jogador = tmp;
                i = i.prox;
                j = j.ant;
            }
        }
        if (esq != j && esq.ant != j)  quicksort(esq, j);
        if (i != dir && i!= dir.prox)  quicksort(i, dir);
    } 
      static Celula getPivo(Celula esq, Celula dir)
{
	while(esq != dir && esq.prox!=dir)
	{
		esq = esq.prox;
		dir = dir.ant;
	}
	return esq;
}
   static int compare2(Jogador a, Jogador b)
{
	String jog1;
	String jog2;
	jog1 = a.getEstadoNascimento() + a.getNome().trim();
	jog2 = b.getEstadoNascimento() + b.getNome().trim();
	return jog1.compareTo(jog2);
} 
  public static void main(String args[]) throws Exception 
  {
	   String temp;
	   Jogador[] a = ler();
	   int i = 0;
	   ListaFlex lf = new ListaFlex();
	    while(true)
	    {
		   temp =  MyIO.readLine();
		   if(temp.equals("FIM"))
				   break;
                   else if(Integer.parseInt(temp) <3922)
		   {
			   lf.inserirFim(PesquisaId(Integer.parseInt(temp)));
			   i++;
		   }
             }
	            lf.sort();
		    lf.mostrar();
		    

  
  }
}

